executor/executor.go

```diff
@@ -41,118 +50,108 @@ func New(items, concurrency int, metrics Metrics) *Executor {
 		metrics:    metrics,
 		stop:       make(chan struct{}),
 		tasks:      make(map[int]*task, items),
-		edges:      make(map[string]int, items*2), // TODO: tune this
-		executable: make(chan *task, items),       // ensure we don't block while holding lock
+		edges:      make(map[string]*data, items*2), // TODO: tune this
+		executable: make(chan *task, items),         // ensure we don't block while holding lock
 	}
+	e.wg.Add(concurrency)
 	for i := 0; i < concurrency; i++ {
-		e.createWorker()
+		go e.runWorker()
 	}
 	return e
 }
 
 type task struct {
-	id int
-	f  func() error
-
-	dependencies set.Set[int]
-	blocking     set.Set[int]
-
-	executed bool
+	id   int
+	f    func() error
+	keys state.Keys
 }
 
-func (e *Executor) createWorker() {
-	e.wg.Add(1)
+func (e *Executor) runWorker() {
+	defer e.wg.Done()
 
-	go func() {
-		defer e.wg.Done()
+	for {
+		select {
+		case t, ok := <-e.executable:
+			if !ok {
+				return
+			}
 
-		for {
-			select {
-			case t, ok := <-e.executable:
-				if !ok {
-					return
-				}
-				if err := t.f(); err != nil {
-					e.stopOnce.Do(func() {
-						e.err = err
-						close(e.stop)
-					})
-					return
-				}
+			if err := t.f(); err != nil {
+				e.stopOnce.Do(func() {
+					e.err = err
+					close(e.stop)
+				})
+				return
+			}
 
-				e.l.Lock()
-				for b := range t.blocking { // works fine on non-initialized map
-					bt := e.tasks[b]
-					bt.dependencies.Remove(t.id)
-					if bt.dependencies.Len() == 0 { // must be non-nil to be blocked
-						bt.dependencies = nil // free memory
-						e.executable <- bt
-					}
-				}
-				t.blocking = nil // free memory
-				t.executed = true
-				e.completed++
-				if e.done && e.completed == len(e.tasks) {
-					// We will close here if there are unexecuted tasks
-					// when we call [Wait].
-					close(e.executable)
+			e.l.Lock()
+			for k := range t.keys {
+				key := e.edges[k]
+				key.blockers--
+				if key.blockers == 0 {
+					close(key.waiter)
+					// Allows us to always make a [waiter] chan
+					// after executing the last blocked task.
+					delete(e.edges, k)
 				}
-				e.l.Unlock()
-			case <-e.stop:
-				return
 			}
+			e.completed++
+			if e.done && e.completed == len(e.tasks) {
+				// We will close here if there are unexecuted tasks
+				// when we call [Wait].
+				close(e.executable)
+			}
+			e.l.Unlock()
+		case <-e.stop:
+			return
 		}
-	}()
+	}
 }
 
 // Run executes [f] after all previously enqueued [f] with
 // overlapping [conflicts] are executed.
-// TODO: Handle read-only/write-only keys (currently the executor
-// treats everything still as ReadWrite, see issue below)
-// https://github.com/ava-labs/hypersdk/issues/709
 func (e *Executor) Run(conflicts state.Keys, f func() error) {
 	e.l.Lock()
```

Comment:
From: https://github.com/ava-labs/hypersdk/blob/3c56945e64f0b76415b27bac30341c10772644bb/executor/executor.go 

I think this can applied here as well